### PR TITLE
Support field parameters

### LINKED ISSUE #272: How to pass variables to fields?
Considering following GraphQL query, how can I query the same with mst-gql's autogenerated root queries?
```graphql
{
  user {
    project(name:"MobX") {
      name
    }
  }
}
```

```ts
rootStore.queryUser({}, user => user
  // How can I pass the `name` variable here?
  .project(project => project.name)
);
```
Thanks!

### PR DESCRIPTION
**Pull request by smokku**
Adds support for parameterized fields.

Fixes #272

### CODE REVIEW

**Comment by beepsoft** [comment]
@jesse-savary do you plan to merge this PR (and the others) if you have the time? Would be great to see some progress of the project.

**Comment by jesse-savary** [comment]
Yes @beepsoft -- going forward this next quarter I've got time budgeted in each sprint to work on this project. Apologies for the delays with things and I appreciate the ping.

**Comment by beepsoft** [comment]
Thanks, @jesse-savary, really appreciate your work on the project!

**Comment by beepsoft** [comment]
Will you make a new release with the merged PR-s?

**Comment by beepsoft** [comment]
Hi @jesse-savary! Do you have an estimate for a new release?

**Comment by jesse-savary** [comment]
> Hi @jesse-savary! Do you have an estimate for a new release?

Hey, I just need to fix up a few tests and we're good to go. I will do my best to solve them tomorrow.

**Comment by jesse-savary** [comment]
Released as `0.15.0` @smokku @beepsoft

**Comment by beepsoft** [comment]
@jesse-savary, I cannot find it on [npm](https://www.npmjs.com/), it still has only [0.14.1](https://www.npmjs.com/package/mst-gql/v/0.14.1)

https://www.npmjs.com/package/mst-gql

**Comment by jesse-savary** [comment]
@beepsoft made the mistake of publishing it to my private registry only; fixed now

**Comment by beepsoft** [comment]
I just tried 0.15.0 with a Hasura generated schema and there seems to be some problems. I have a test project for it:

https://github.com/beepsoft/mst-gql-0.15.0-app

For example, in src/PollsModel.base.ts the imports for ChoicesOrderBy, ChoicesBoolExp, etc are not generated. But there are other TS errors as well.

It may worth considering adding a test using this somewhat complex Hasura schema (https://gist.github.com/beepsoft/84a64102a80a3ac7eb1f656b64fc2af8) to see whether generation succeeds with it.

**Comment by beepsoft** [comment]
I just added a failing test in #360 to help figure out the bug.

**Comment by beepsoft** [comment]
In the meantime I tried to fix it here #361.

On thing that bothers me from devex point of view is that the signature of the selector functions is

`function field(builder, args)`

which makes it unintuitive because in graphql it will generate

```
   field(args) {
     <selectors built by builder>
   }
```

For example:

```typescript
store.queryTasks(
    {},
    selectFromWorkInstruction().operations((builder) =>
        builder
            .operationId
            .instructionId
            .operationOrder,
        {
            limit: 2
        }
    ).toString()
)
```

will generate:

``` graphql
query {
  tasks {
    operations(limit: 2) {
      operationId
      instructionId
      operationOrder
    }
  }
}
```

So, when defining selectors using the generated builder functions all the sub fields will appear first in the code and the field arguments at the end. However it is not easy to switch the `builder` and `args` params, since `args` can be optional and builder is required.

One solution might be to change the selector function to have a single object parameter with builder and args fields:

`function field({builder, args})`

In this case the caller can decide which order to have builder and args, with the tradeoff of some verboseness needed to write out `args` and `builder`

```typescript
store.queryTasks(
    {},
    selectFromWorkInstruction().operations({
      args: {
          limit: 2
      },
      builder: (builder) =>
          builder
            .operationId
            .instructionId
            .operationOrder,
    }).toString()
```

Any other ideas?

File: generator/generate.js [hunk 1]
```diff
@@ -536,13 +554,29 @@ ${generateFragments(name, primitiveFields, nonPrimitiveFields)}
...
     output += primitiveFields.length > 0 ? "\n" : ""
     output += nonPrimitiveFields
-      .map(([field, fieldName]) => {
+      .map(([field, fieldName, fieldArgs]) => {
+        let args = ""
         const selector = `${fieldName}ModelSelector`
         let p = `  ${field}(builder`
         p += ifTS(
-          `?: string | ${selector} | ((selector: ${selector}) => ${selector})`
+          `: string | ${selector} | ((selector: ${selector}) => ${selector}) | undefined`
```
Comment (wtuminski) [inline generator/generate.js:562]: @jesse-savary, is it necessary to drop the optional parameter? If so it expects to pass `undefined` as a parameter each time we want to get a child with only __typename and id.

File: generator/generate.js [hunk 2]
```diff
@@ -536,13 +554,29 @@ ${generateFragments(name, primitiveFields, nonPrimitiveFields)}
...
+          p += ", args"
+          p += ifTS(
+            `${required ? "" : "?"}: { ${fieldArgs
+              .map((arg) => `${printTsType(arg)}`)
+              .join(", ")} }`
+          )
+          args = `(${fieldArgs
+            .map(({ name }) => `${name}: \${JSON.stringify(args['${name}'])}`)
+            .join(", ")})`
+          console.log(fieldArgs)
```
Comment (wtuminski) [inline generator/generate.js:577]: Debug code

Comment (beepsoft) [reply generator/generate.js:577]: PR #361 fixes it, already merged to main.

Comment (wtuminski) [reply generator/generate.js:577]: great!

**Pull request merged**